### PR TITLE
Click and drag to move selection in pattern editor

### PR DESCRIPTION
**Pull request by exelotl**
Hi! I've been working on a feature that I couldn't live without (after using Renoise for 6 years), I hope you like it!

GIF 2020-05-03 02-30-05

Summary of changes:
- Click and drag a selection in the pattern editor to move the selected pattern data
  - Use the modifier key to clone instead of moving
- More intuitive behaviour for cut/paste in the effect column
  - When cutting an effect type, the parameter value is preserved
  - When pasting an effect parameter value into an empty slot, the effect type will be set to Arpeggio (previously the effect would remain as "No Effect" so the value would not be pasted)
- Improved mouse handling near the edges of the pattern (selection is clamped to boundaries instead of ignoring input)
- Added MSYS2 + MinGW support to the CMake configuration

### CODE REVIEW

**Comment by dwhinham** [comment]
Thanks for the PR, the new functionality looks cool!

The MSYS2/MinGW stuff is kindof irrelevant though; I'm not sure we need this when we build for Windows with its native toolchain.

I'd be inclined to squash the drag-select stuff into one commit and merge that, then maybe think about why we might want to support another toolchain.

File: src/tracker/PatternEditorTools.cpp [hunk 1]
```diff
@@ -1718,6 +1718,40 @@ void PatternEditorTools::normalize()
...
+	pp_int32 selectionEndRow;
+	pp_int32 selectionEndInner;
+	
+	if (!normalizeSelection(pattern, ss, se, 
+							selectionStartChannel, selectionStartRow, selectionStartInner,
+							selectionEndChannel, selectionEndRow,selectionEndInner))
+		return false;
+	
+	// only entire instrument column is allowed
+	if (selectionStartInner >= 1 && selectionStartInner<=2)
```
Comment (dwhinham) [inline src/tracker/PatternEditorTools.cpp:1736]: Nitpick: let's try to keep spacing consistent.
`selectionEndInner >= 1 && selectionEndInner<=2`
becomes
`selectionEndInner >= 1 && selectionEndInner <= 2`

File: src/tracker/PatternEditorTools.cpp [hunk 2]
```diff
@@ -1888,4 +1922,111 @@ void PatternEditorTools::slotClear(mp_ubyte* dst, pp_int32 from, pp_int32 to)
...
+	pp_int32 selectionEndRow;
+	pp_int32 selectionEndInner;
 
+	if (!normalizeSelection(pattern, ss, se, 
+							selectionStartChannel, selectionStartRow, selectionStartInner,
+							selectionEndChannel, selectionEndRow,selectionEndInner))
+		return 0;
+	
+	// only entire instrument column is allowed
+	if (selectionStartInner >= 1 && selectionStartInner<=2)
```
Comment (dwhinham) [inline src/tracker/PatternEditorTools.cpp:1991]: Same here re: spacing

File: src/tracker/PatternEditorTools.cpp [hunk 3]
```diff
@@ -1888,4 +1922,111 @@ void PatternEditorTools::slotClear(mp_ubyte* dst, pp_int32 from, pp_int32 to)
 	i++;
 }
 
+bool PatternEditorTools::moveSelection(const Position& ss, const Position& se, pp_int32 moveChannels, pp_int32 moveRows, bool clear)
+{
+	if (!PatternEditorTools::hasValidSelection(pattern, ss, se))
+		return 0;
```
Comment (dwhinham) [inline src/tracker/PatternEditorTools.cpp:1974]: Let's return `true` or `false` rather than 0/1

File: src/tracker/PatternEditorTools.cpp [hunk 4]
```diff
@@ -1888,4 +1922,111 @@ void PatternEditorTools::slotClear(mp_ubyte* dst, pp_int32 from, pp_int32 to)
...
+	if (selectionStartInner >= 3 && selectionStartInner<=4)
+		selectionStartInner = 3;
+	if (selectionEndInner >= 3 && selectionEndInner<=4)
+		selectionEndInner = 4;
+	
+	pp_int32 selectionWidth = selectionEndChannel - selectionStartChannel + 1;
+	pp_int32 selectionHeight = selectionEndRow - selectionStartRow + 1;
+	
+	mp_sint32 slotSize = pattern->effnum * 2 + 2;
+	ASSERT(slotSize == 6);
```
Comment (dwhinham) [inline src/tracker/PatternEditorTools.cpp:2005]: A comment explaining why we expect `slotSize == 6` would be helpful here

**Comment by exelotl** [comment]
A little busy rn but I'll make the requested changes when I can!

I included the MSYS2/MinGW stuff cause I can't build without it. The changes required to get it working were relatively small. Not expecting the project to claim official support or anything, but I literally don't have space on my machine to install gigabytes of visual studio tools. 😅

Please confirm if you want me to make another PR without any superfluous stuff. I can do it but to me it seems like extra effort for the same outcome. If we don't support MSYS2/MinGW I'm gonna have to rebase or cherry pick my commits every time I want to contribute.

**Comment by Deltafire** [comment]
Side note: 6 years ago we were using MinGW for the Windows release builds, we discovered it was breaking ASIO support:

https://github.com/milkytracker/MilkyTracker/issues/4

Since then we've been building with Visual Studio.

File: src/tracker/PatternEditorTools.cpp (hunk 4 above)
Comment (exelotl) [reply src/tracker/PatternEditorTools.cpp:2007]: This was done in PatternEditorClipBoard so it seemed like I should do the same. The purpose of `effnum` isn't documented anywhere but I might assume it goes like this:

- Milky's internal pattern format is able to represent formats other than XM, for conversion purposes
- The format always has 1 byte for "note" and 1 byte for "instrument number"
- `pattern->effnum` is used to specify how much extra 'effect' data each slot contains (in 2-byte increments)
- When representing XM, this means we use an extra 2 bytes for volume, and an extra 2 bytes for effect commands
- `slotCopy` and friends are only intended to work with the XM representation, not any other format, so we assert `slotSize == 6` as a sanity check.

This is a lot of information to convey, and not the best place for it, but I've left a short comment to explain why the check is there at least.

File: src/tracker/PatternEditor.cpp [hunk 5]
```diff
@@ -211,6 +211,11 @@ bool PatternEditor::hasValidSelection()
 	return PatternEditorTools::hasValidSelection(pattern, selection.start, selection.end, getNumChannels());
 }
 
+bool PatternEditor::selectionContains(const PatternEditorTools::Position& pos)
+{
+	return PatternEditorTools::selectionContains(pattern ,selection.start, selection.end, pos);
```
[the remark this answers is not in the document]
Comment (exelotl) [reply src/tracker/PatternEditor.cpp]: Already fixed in a subsequent commit

File: .gitignore [hunk 6]
```diff
@@ -4,3 +4,6 @@ build/
 # Resources generated with docerator on OSX
 resources/pictures/docicons/osx/MilkyTracker-*.icns
 resources/pictures/docicons/osx/docerator/
+
```
[the remark this answers is not in the document]
Comment (exelotl) [reply .gitignore:7]:
```diff
+ # Editor config files
+ .vscode/
```
Do I need to do anything about this? Doesn't harm anyone 😕

File: src/tracker/PatternEditorControl.cpp [hunk 7]
```diff
@@ -826,30 +826,44 @@ void PatternEditorControl::paint(PPGraphicsAbstract* g)
...
-		const pp_int32 dashLen = 6;
-		
-		// inner dashed lines
-		g->drawHLineDashed(x1, x2, y1, dashLen, 3);
-		g->drawHLineDashed(x1, x2, y2, dashLen, 3);
-		g->drawVLineDashed(y1, y2, x1, dashLen, 3);
-		g->drawVLineDashed(y1, y2+1, x2, dashLen, 3);
+		if (i2 >= 0 && j2 >= 0 && i1 < pattern->rows && j1 < numVisibleChannels)
+		{
+			#define CLAMP(a, min, max) ((a) < (min) ? (min) : ((a) >= (max) ? (max-1) : (a)))
```
[the remark this answers is not in the document]
Comment (exelotl) [reply src/tracker/PatternEditorControl.cpp]: After some digging, moving this to PPTools seemed appropriate. (https://github.com/milkytracker/MilkyTracker/pull/211/commits/e9658cf4fe0bb2c6141ebbe81c6245e233f97c26)

**Comment by exelotl** [comment]
Should be good now. If you're up for merging the pattern editor commits, I'll look into making another PR for the MSYS2 support later if you're happy for that discussion to continue. Sorry for jumping the gun with that stuff.

**Comment by exelotl** [comment]
Would you like me to revert the CMake commits to make it easier to merge?

File: .gitignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwhinham) [reply .gitignore:7]: These are personal preferences. You don't commit personal preferences to a shared project.
You don't know that it doesn't harm others.

No need to fix, we'll just ignore it, but be aware in future.

**Comment by dwhinham** [comment]
Apologies for slow response, things have been busy back in real-life work.

Thanks for making the requested amendments, everything looks good; don't worry about reverting stuff, as mentioned we'll handle it when squashing.

I think we can forget about MSYS to be honest; as @Deltafire mentioned (I completely forgot about this) it apparently has ABI problems that lead to breakage with the ASIO backend.

If you don't want to install all of Visual Studio, you can just install the [standalone MSVC compiler/build tools](https://visualstudio.microsoft.com/visual-cpp-build-tools/). I haven't tested this, but it should work.

**Comment by Deltafire** [comment]
I've removed the MinGW commits, fixedup, squashed, reworded and merged into master.

Thank-you for the PR :)

**Pull request closed without merging**